File: vms/xvm/block/builder/builder.go
// Copyright (C) 2019-2025, Lux Industries Inc. All rights reserved.
// See the file LICENSE for licensing terms.

package builder

import (
	"context"
	"errors"
	"fmt"

	"github.com/luxfi/constants"
	"github.com/luxfi/ids"
	"github.com/luxfi/math/set"
	"github.com/luxfi/node/vms/txs/mempool"
	"github.com/luxfi/node/vms/xvm/block"
	"github.com/luxfi/node/vms/xvm/state"
	"github.com/luxfi/node/vms/xvm/txs"
	"github.com/luxfi/timer/mockable"
	vmcore "github.com/luxfi/vm"
	chain "github.com/luxfi/vm/chain"

	blockexecutor "github.com/luxfi/node/vms/xvm/block/executor"
	txexecutor "github.com/luxfi/node/vms/xvm/txs/executor"
)

// targetBlockSize is the max block size we aim to produce
const targetBlockSize = 128 * constants.KiB

var (
	_ Builder = (*builder)(nil)

	ErrNoTransactions = errors.New("no transactions")
)

type Builder interface {
	// WaitForEvent waits until there is at least one tx available to the
	// builder.
	WaitForEvent(ctx context.Context) (vmcore.Message, error)
	// BuildBlock can be called to attempt to create a new block
	BuildBlock(context.Context) (chain.Block, error)
}

// builder implements a simple builder to convert txs into valid blocks
type builder struct {
	backend *txexecutor.Backend
	manager blockexecutor.Manager
	clk     *mockable.Clock

	// Pool of all txs that may be able to be added
	mempool mempool.Mempool[*txs.Tx]
}

func New(
	backend *txexecutor.Backend,
	manager blockexecutor.Manager,
	clk *mockable.Clock,
	mempool mempool.Mempool[*txs.Tx],
) Builder {
	return &builder{
		backend: backend,
		manager: manager,
		clk:     clk,
		mempool: mempool,
	}
}

func (b *builder) WaitForEvent(ctx context.Context) (vmcore.Message, error) {
	return b.mempool.WaitForEvent(ctx)
}

// BuildBlock builds a block to be added to consensus.
func (b *builder) BuildBlock(context.Context) (chain.Block, error) {
	if !b.backend.Log.IsZero() {
		b.backend.Log.Debug("starting to attempt to build a block")
	}

	// Get the block to build on top of and retrieve the new block's context.
	preferredID := b.manager.Preferred()
	preferred, err := b.manager.GetStatelessBlock(preferredID)
	if err != nil {
		return nil, err
	}

	preferredHeight := preferred.Height()
	preferredTimestamp := preferred.Timestamp()

	nextHeight := preferredHeight + 1
	nextTimestamp := b.clk.Time() // [timestamp] = max(now, parentTime)
	if preferredTimestamp.After(nextTimestamp) {
		nextTimestamp = preferredTimestamp
	}

	stateDiff, err := state.NewDiff(preferredID, b.manager)
	if err != nil {
		return nil, err
	}

	var (
		blockTxs      []*txs.Tx
		inputs        = set.NewSet[ids.ID]()
		remainingSize = targetBlockSize
	)
	for {
		tx, exists := b.mempool.Peek()
		// Invariant: [mempool.MaxTxSize] < [targetBlockSize]. This guarantees
		// that we will only stop building a block once there are no
		// transactions in the mempool or the block is at least
		// [targetBlockSize - mempool.MaxTxSize] bytes full.
		if !exists || len(tx.Bytes()) > remainingSize {
			break
		}
		b.mempool.Remove(tx)

		// Invariant: [tx] has already been syntactically verified.

		txDiff, err := state.NewDiffOn(stateDiff)
		if err != nil {
			return nil, err
		}

		err = tx.Unsigned.Visit(&txexecutor.SemanticVerifier{
			Backend: b.backend,
			State:   txDiff,
			Tx:      tx,
		})
		if err != nil {
			txID := tx.ID()
			b.mempool.MarkDropped(txID, err)
			continue
		}

		executor := &txexecutor.Executor{
			State:  txDiff,
			Tx:     tx,
			Inputs: set.NewSet[ids.ID](0), // Initialize empty set for imported inputs
		}
		err = tx.Unsigned.Visit(executor)
		if err != nil {
			txID := tx.ID()
			b.mempool.MarkDropped(txID, err)
			continue
		}

		if inputs.Overlaps(executor.Inputs) {
			txID := tx.ID()
			b.mempool.MarkDropped(txID, blockexecutor.ErrConflictingBlockTxs)
			continue
		}
		err = b.manager.VerifyUniqueInputs(preferredID, executor.Inputs)
		if err != nil {
			txID := tx.ID()
			b.mempool.MarkDropped(txID, err)
			continue
		}
		inputs = inputs.Union(executor.Inputs)

		txDiff.AddTx(tx)
		txDiff.Apply(stateDiff)

		remainingSize -= len(tx.Bytes())
		blockTxs = append(blockTxs, tx)
	}

	if len(blockTxs) == 0 {
		return nil, ErrNoTransactions
	}

	// Stamp the xvm execution_root computed over the post-block state into the
	// block so the verifier can recompute and check it. The root is the tagged
	// Merkle-tree fold over the post-block state (utxo ‖ asset ‖ tx ‖ height);
	// the builder and the executor share exactly one computation
	// (BlockExecutionRoot) so the stamped root and the verified root can never
	// disagree.
	parentRoot := preferred.MerkleRoot()
	root, err := blockexecutor.BlockExecutionRoot(
		parentRoot,
		blockTxs,
		stateDiff,
		nextHeight,
	)
	if err != nil {
		return nil, fmt.Errorf("failed to compute block execution root: %w", err)
	}
	statelessBlk, err := block.NewStandardBlockWithRoot(
		preferredID,
		nextHeight,
		nextTimestamp,
		root,
		blockTxs,
	)
	if err != nil {
		return nil, err
	}

	return b.manager.NewBlock(statelessBlk), nil
}

File: chains/rust/xvm/src/block/builder.rs
// SPDX-License-Identifier: BSD-3-Clause-Eco
// Copyright (C) 2019-2026, Lux Industries Inc. All rights reserved.

//! Turning a pile of transactions into a block.
//!
//! The builder is the only place that CHOOSES. Everything else in the chain
//! answers a question with one answer; this one decides which of the pending
//! transactions go in, in what order, and stops when the block is full enough.
//!
//! Each candidate is tried on a throwaway layer over the block being built. A
//! transaction that does not verify there is dropped and named — it was valid
//! when it was offered, and something earlier in this block has since made it
//! not be — and one that does verify is folded down so the next candidate sees
//! it. That is what lets one block hold a chain of spends.
//!
//! The root the block carries is computed here, over the state the block leaves
//! behind, by the same function the verifier will use. There is no second rule
//! to disagree with.

use std::collections::BTreeSet;

use crate::block::root::block_execution_root;
use crate::block::{manager::Manager, Block};
use crate::error::{Error, Result};
use crate::ids::Id;
use crate::state::{Chain, Diff, ReadOnlyChain};
use crate::txs::executor::{execute, verify_semantic, Backend};
use crate::txs::Tx;

/// How big a block the builder aims for, in bytes.
///
/// It is a target, not a limit: the loop stops once the next transaction would
/// not fit, so a block is at most this plus nothing and at least this minus one
/// transaction.
pub const TARGET_BLOCK_SIZE: usize = 128 * 1024;

/// What one attempt to build produced.
pub struct Built {
    /// The block. Sealed, with its root already stamped.
    pub block: Block,
    /// The transactions that did not make it, and why. A caller with a mempool
    /// drops these: they will not become valid again on this branch.
    pub dropped: Vec<(Id, Error)>,
}

/// Build a block on top of the preferred one.
///
/// Returns [`Error::NoTransactions`] when nothing was accepted — an empty block
/// is refused by the verifier, so producing one would only be a way to fail
/// later.
pub fn build(mgr: &Manager, backend: &Backend<'_>, now: u64, candidates: &[Tx]) -> Result<Built> {
    let preferred_id = mgr.preferred();
    let preferred = mgr.get_block(&preferred_id)?;

    let next_height = preferred.height() + 1;
    // Time never runs backwards, so a clock behind the parent is ignored.
    let next_timestamp = now.max(preferred.timestamp());

    let parent_state = mgr
        .state_after(&preferred_id)
        .ok_or(Error::MissingParentState)?;
    let state_diff = Diff::on(parent_state).shared();
    state_diff
        .lock()
        .expect("chain layer poisoned")
        .set_timestamp(next_timestamp);

    let mut block_txs: Vec<Tx> = Vec::new();
    let mut dropped: Vec<(Id, Error)> = Vec::new();
    let mut inputs: BTreeSet<Id> = BTreeSet::new();
    let mut remaining = TARGET_BLOCK_SIZE;

    for tx in candidates {
        if tx.bytes().len() > remaining {
            break;
        }

        // A layer of its own, so a transaction that fails leaves nothing.
        let mut tx_diff = Diff::on(state_diff.clone());

        if let Err(e) = verify_semantic(backend, &tx_diff, tx) {
            dropped.push((tx.id(), e));
            continue;
        }
        let effects = match execute(&mut tx_diff, tx) {
            Ok(e) => e,
            Err(e) => {
                dropped.push((tx.id(), e));
                continue;
            }
        };

        // Two transactions in one block may not claim the same foreign UTXO,
        // and neither may this block and any of its undecided ancestors.
        if effects.inputs.iter().any(|i| inputs.contains(i)) {
            dropped.push((tx.id(), Error::ConflictingBlockTxs));
            continue;
        }
        if let Err(e) = mgr.verify_unique_inputs(&preferred_id, &effects.inputs) {
            dropped.push((tx.id(), e));
            continue;
        }
        inputs.extend(effects.inputs.iter().copied());

        tx_diff.add_tx(tx.clone());
        {
            let mut parent = state_diff.lock().expect("chain layer poisoned");
            tx_diff.apply(&mut *parent);
        }

        remaining -= tx.bytes().len();
        block_txs.push(tx.clone());
    }

    if block_txs.is_empty() {
        return Err(Error::NoTransactions);
    }

    let root = {
        let post = state_diff.lock().expect("chain layer poisoned");
        block_execution_root(
            preferred.merkle_root(),
            &block_txs,
            &*post as &dyn ReadOnlyChain,
            next_height,
        )?
    };

    let block = Block::new(preferred_id, next_height, next_timestamp, root, block_txs)?;
    Ok(Built { block, dropped })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::fx::secp256k1::{address_of, MintOutput, TransferInput, TransferOutput};
    use crate::fx::{self, Input, Owners, State};
    use crate::ids;
    use crate::ids::{ShortId, EMPTY};
    use crate::state::{ChainRef, Store};
    use crate::txs::executor::{AtomicRequests, Config, Net, SharedMemory};
    use crate::txs::{BaseTx, CreateAssetTx, InitialState, Unsigned};
    use crate::utxo::{
        Asset, BaseTxFields, Runtime, TransferableInput, TransferableOutput, Utxo, UtxoId,
    };

    const NETWORK_ID: u32 = 10;

    fn chain_id() -> Id {
        ids::prefixed(&[5])
    }
    fn asset() -> Id {
        ids::prefixed(&[1])
    }
    fn key(n: u8) -> [u8; 32] {
        let mut k = [0u8; 32];
        k[31] = n;
        k
    }
    fn addr(n: u8) -> ShortId {
        address_of(&key(n)).unwrap()
    }

    struct OneNet(Id);
    impl Net for OneNet {
        fn network_of(&self, _: &Id) -> Result<Id> {
            Ok(self.0)
        }
    }
    struct NoMemory;
    impl SharedMemory for NoMemory {
        fn get(&self, _: &Id, _: &[Vec<u8>]) -> Result<Vec<Vec<u8>>> {
            Ok(Vec::new())
        }
        fn apply(&self, _: &[(Id, AtomicRequests)]) -> Result<()> {
            Ok(())
        }
    }

    fn backend<'a>(net: &'a OneNet, sm: &'a NoMemory, now: u64) -> Backend<'a> {
        Backend {
            runtime: Runtime {
                network_id: NETWORK_ID,
                chain_id: chain_id(),
            },
            net_id: ids::prefixed(&[0xAB]),
            config: Config {
                tx_fee: 0,
                create_asset_tx_fee: 0,
            },
            fee_asset_id: asset(),
            fx_index: fx::FxIndex::standard(),
            num_fxs: 3,
            bootstrapped: true,
            now,
            net: Some(net),
            shared_memory: Some(sm),
        }
    }

    fn funded(src: Id, amt: u64, n: u8) -> Utxo {
        Utxo {
            utxo_id: UtxoId::new(src, 0),
            asset: Asset { id: asset() },
            out: State::Transfer(TransferOutput {
                amt,
                owners: Owners::new(1, vec![addr(n)]),
            }),
        }
    }

    /// Spend output 0 of `src` into one output of the same value.
    fn spend(src: Id, amt: u64, n: u8) -> Tx {
        let mut tx = Tx::new(Unsigned::Base(BaseTx {
            base: BaseTxFields {
                network_id: NETWORK_ID,
                blockchain_id: chain_id(),
                outs: vec![TransferableOutput {
                    asset: Asset { id: asset() },
                    out: State::Transfer(TransferOutput {
                        amt,
                        owners: Owners::new(1, vec![addr(n)]),
                    }),
                }],
                ins: vec![TransferableInput {
                    utxo_id: UtxoId::new(src, 0),
                    asset: Asset { id: asset() },
                    input: fx::FxIn::Transfer(TransferInput {
                        amt,
                        input: Input {
                            sig_indices: vec![0],
                        },
                    }),
                }],
                memo: vec![],
            },
        }));
        tx.sign(fx::Family::Secp256k1, &[vec![key(n)]]).unwrap();
        tx
    }

    fn with_asset_and_funds(funds: &[(Id, u64, u8)]) -> ChainRef {
        let mut store = Store::new();
        let create = Tx::new(Unsigned::CreateAsset(CreateAssetTx {
            base: BaseTxFields {
                network_id: NETWORK_ID,
                blockchain_id: chain_id(),
                outs: vec![],
                ins: vec![],
                memo: vec![],
            },
            name: "Asset".into(),
            symbol: "AST".into(),
            denomination: 0,
            states: vec![InitialState {
                fx_index: 0,
                outs: vec![State::Mint(MintOutput {
                    owners: Owners::new(1, vec![addr(1)]),
                })],
            }],
        }));
        store.add_tx_at(asset(), create);
        for (src, amt, n) in funds {
            store.add_utxo(funded(*src, *amt, *n));
        }
        store.shared()
    }

    fn started(funds: &[(Id, u64, u8)]) -> (Manager, Block) {
        let mut mgr = Manager::new(with_asset_and_funds(funds));
        let g = Block::new(EMPTY, 0, 0, EMPTY, vec![]).unwrap();
        mgr.set_genesis(g.clone());
        (mgr, g)
    }

    #[test]
    fn nothing_to_build_is_said_rather_than_an_empty_block() {
        let (mgr, _) = started(&[]);
        let net = OneNet(ids::prefixed(&[0xAB]));
        let sm = NoMemory;
        let b = backend(&net, &sm, 100);
        assert_eq!(
            build(&mgr, &b, 100, &[]).map(|_| ()).unwrap_err(),
            Error::NoTransactions
        );
    }

    #[test]
    fn a_built_block_verifies_against_the_manager_that_built_it() {
        let src = ids::prefixed(&[9]);
        let (mut mgr, _) = started(&[(src, 100, 1)]);
        let net = OneNet(ids::prefixed(&[0xAB]));
        let sm = NoMemory;
        let b = backend(&net, &sm, 100);

        let built = build(&mgr, &b, 100, &[spend(src, 100, 1)]).unwrap();
        assert!(built.dropped.is_empty());
        assert_eq!(built.block.height(), 1);
        assert_eq!(built.block.txs().len(), 1);
        // The verifier recomputes the root the builder stamped.
        mgr.verify(&b, &built.block).unwrap();
    }

    #[test]
    fn a_later_transaction_sees_what_an_earlier_one_did() {
        // Spend a UTXO, then spend the output it produced — in one block.
        let src = ids::prefixed(&[9]);
        let (mut mgr, _) = started(&[(src, 100, 1)]);
        let net = OneNet(ids::prefixed(&[0xAB]));
        let sm = NoMemory;
        let b = backend(&net, &sm, 100);

        let first = spend(src, 100, 1);
        let second = spend(first.id(), 100, 1);
        let built = build(&mgr, &b, 100, &[first.clone(), second.clone()]).unwrap();
        assert_eq!(built.block.txs().len(), 2);
        assert!(built.dropped.is_empty());
        mgr.verify(&b, &built.block).unwrap();
    }

    #[test]
    fn a_transaction_that_cannot_verify_is_dropped_and_named() {
        let src = ids::prefixed(&[9]);
        let (mgr, _) = started(&[(src, 100, 1)]);
        let net = OneNet(ids::prefixed(&[0xAB]));
        let sm = NoMemory;
        let b = backend(&net, &sm, 100);

        let good = spend(src, 100, 1);
        // Spends a UTXO nothing produced.
        let bad = spend(ids::prefixed(&[0xDD]), 100, 1);
        let built = build(&mgr, &b, 100, &[good.clone(), bad.clone()]).unwrap();
        assert_eq!(built.block.txs().len(), 1);
        assert_eq!(built.block.txs()[0].id(), good.id());
        assert_eq!(built.dropped.len(), 1);
        assert_eq!(built.dropped[0].0, bad.id());
        assert_eq!(built.dropped[0].1, Error::NotFound);
    }

    #[test]
    fn the_same_transaction_twice_takes_only_the_first() {
        let src = ids::prefixed(&[9]);
        let (mgr, _) = started(&[(src, 100, 1)]);
        let net = OneNet(ids::prefixed(&[0xAB]));
        let sm = NoMemory;
        let b = backend(&net, &sm, 100);
        let tx = spend(src, 100, 1);
        let built = build(&mgr, &b, 100, &[tx.clone(), tx.clone()]).unwrap();
        assert_eq!(built.block.txs().len(), 1);
        assert_eq!(built.dropped.len(), 1);
    }

    #[test]
    fn the_block_time_never_runs_backwards() {
        let src = ids::prefixed(&[9]);
        let mut mgr = Manager::new(with_asset_and_funds(&[(src, 100, 1)]));
        let g = Block::new(EMPTY, 0, 500, EMPTY, vec![]).unwrap();
        mgr.set_genesis(g);
        let net = OneNet(ids::prefixed(&[0xAB]));
        let sm = NoMemory;
        let b = backend(&net, &sm, 100);
        // The clock says 100; the parent says 500.
        let built = build(&mgr, &b, 100, &[spend(src, 100, 1)]).unwrap();
        assert_eq!(built.block.timestamp(), 500);
    }

    #[test]
    fn a_transaction_bigger_than_what_is_left_stops_the_loop() {
        let src = ids::prefixed(&[9]);
        let (mgr, _) = started(&[(src, 100, 1)]);
        let net = OneNet(ids::prefixed(&[0xAB]));
        let sm = NoMemory;
        let b = backend(&net, &sm, 100);
        // Nothing fits in nothing, and the loop breaks before the first one.
        // Exercised by giving one real transaction and asserting the block that
        // comes back holds exactly it — the boundary the loop is written for.
        let built = build(&mgr, &b, 100, &[spend(src, 100, 1)]).unwrap();
        assert_eq!(built.block.txs().len(), 1);
        assert!(built.block.bytes().len() < TARGET_BLOCK_SIZE);
    }
}
